chapters/06.OptimalTransport/optimal_transportation.jl

```diff
@@ -0,0 +1,1792 @@
+2. Complete the function `monge_brute_force` to use brute-force search for the best permutation. You might find the function `permutations` from the `Combinatorics` library useful. How large is your search space? Time your function.
+3. Make a plot connecting the cells.
+"""
+
+# ╔═╡ b85b69fb-bfdb-41bd-a6a1-1aebe3077e08
+with_terminal() do
+	for p in permutations([1, 2, 3])
+  		println(p)
+	end
+end
+
+# ╔═╡ cc62b7b1-fe34-4a19-846d-ca6c7fb69ff9
+function monge_brute_force(C)
+  n, m = size(C)
+  @assert n == m "C should be square"
+  # loop over all permutations and to find the
+  # matching with the lowest cost
+  return best_perm, best_cost
+end
+
+# ╔═╡ 1468ec48-bedb-4977-8b88-794d89e5ae8f
+C_cells = missing
+
+# ╔═╡ 88569ef5-bc60-4a77-aa72-fc9bdf66b3a4
+# add the lines to the cells
+
+# ╔═╡ da2cc178-64c7-4b30-909a-24fb402a489b
+md"""
+## The optimal transportion problem
+
+Let us introduce some notation so we can formally state this as an optimization problem. Let $\mathbf{a}$ be the vector containing the amount of dessert every person may eat. In this case $\mathbf{a} = [3,3,3,4,2,2,2,1]^\intercal$ (in general the dimension of $\mathbf{r}$ is $n$). Similarly, $\mathbf{c}$ denotes the vector of how much there is of every dessert, i.e. $\mathbf{b}=[4, 2, 6, 4, 4]^\intercal$ (in general the dimension of $\mathbf{b}$ is $m$). Often $\mathbf{a}$ and $\mathbf{b}$ represent marginal probability distributions, hence their values are nonzero sum to one.
+
+Let $U(\mathbf{a}, \mathbf{b})$ be the set of positive $n\times m$ matrices for which the rows sum to $\mathbf{a}$ and the columns sum to $\mathbf{b}$:
+
+$$U(\mathbf{a}, \mathbf{b}) = \{P\in \mathbb{R}_{>0}^{n\times m}\mid P\mathbf{1}_m = \mathbf{a}, P^\intercal\mathbf{1}_n = \mathbf{b}\}\,.$$
+
+For our problem, $U(\mathbf{a}, \mathbf{b})$ contains all the ways of dividing the desserts for my colleagues. Note that we assume here that we can slice every dessert however we like. We do not have only to give whole pieces of pie but can provide any fraction we want.
+
+The preferences of each person for each dessert is also stored in a matrix. To be consistent with the literature, this will be stored in a $n\times m$ *cost* matrix $C$. The above matrix is a preference matrix that can easily be changed into a cost matrix by flipping the sign of every element.
+
+So finally, the problem we want to solve is formally posed as
+
+$$d_C(\mathbf{r}, \mathbf{c}) = \min_{P\in U(\mathbf{a}, \mathbf{b})}\, \sum_{i,j}P_{ij}C_{ij}\,.$$
+
+This is called the *optimal transportation* between $\mathbf{a}$ and $\mathbf{b}$. It can be solved relatively easily using linear programming.
+
+The optimum, $d_C(\mathbf{a}, \mathbf{b})$, is called the *Wasserstein metric*. It is basically a distance between two probability distributions. It is sometimes also called the *earth mover distance* as it can be interpreted as how much 'dirt' you have to move to change one 'landscape' (distribution) in another.
```

Comment:
It is basically **the minimum distance** between two probability distributions ?

Reply:
Sentence was correct!